File: xion-ios/XionDemo/Data/Repository/XionRepositoryImpl.swift
import Foundation
import Mob


final class XionRepositoryImpl: XionRepositoryProtocol {

    let sessionManager: SessionManager
    private let mobService: MobSigningServiceProtocol

    init(sessionManager: SessionManager, mobService: MobSigningServiceProtocol) {
        self.sessionManager = sessionManager
        self.mobService = mobService
    }

    func connect() async throws -> String {
        try await sessionManager.authenticate()
    }

    func restoreSession() async -> Bool {
        await sessionManager.restoreSession()
    }

    func getBalance() async throws -> BalanceInfo {
        guard let address = sessionManager.walletState.metaAccountAddress else {
            throw RepositoryError.notConnected
        }
        return try await mobService.getBalance(address: address, denom: Constants.coinDenom)
    }

    func getSbcBalance() async throws -> BalanceInfo {
        guard let address = sessionManager.walletState.metaAccountAddress else {
            throw RepositoryError.notConnected
        }
        return try await mobService.getBalance(address: address, denom: Constants.braleSbcOnChainDenom)
    }

    func getBlockHeight() async throws -> Int64 {
        try await mobService.getHeight()
    }

    func send(toAddress: String, amount: String, memo: String, denom: String = Constants.coinDenom) async throws -> TransactionResult {
        let state = sessionManager.walletState
        guard state.isConnected else { throw RepositoryError.notConnected }

        let coins = [Coin(denom: denom, amount: amount)]
        let result = try await withGrantRecovery {
            try await self.mobService.send(
                toAddress: toAddress,
                coins: coins,
                memo: memo.isEmpty ? nil : memo
            )
        }
        let confirmed = result.success ? (await awaitTxConfirmation(txHash: result.txHash) ?? result) : result
        sessionManager.appendTransaction(confirmed)
        return confirmed
    }

    func executeContract(contractAddress: String, msg: String, funds: String?) async throws -> TransactionResult {
        let state = sessionManager.walletState
        guard state.isConnected else { throw RepositoryError.notConnected }

        let fundsCoins: [Coin] = funds.map { [Coin(denom: Constants.coinDenom, amount: $0)] } ?? []
        let msgData = Data(msg.utf8)
        let result = try await withGrantRecovery {
            try await self.mobService.executeContract(
                contractAddress: contractAddress,
                msg: msgData,
                funds: fundsCoins,
                memo: nil,
                gasLimit: Constants.defaultExecuteGasLimit
            )
        }
        let confirmed = result.success ? (await awaitTxConfirmation(txHash: result.txHash) ?? result) : result
        sessionManager.appendTransaction(confirmed)
        return confirmed
    }

    func getTx(txHash: String) async throws -> TransactionResult {
        try await mobService.getTx(txHash: txHash)
    }

    func getRecentTransactions(address: String) async throws -> [TransactionResult] {
        async let sent = fetchTxsByQuery(address: address, queryKey: "transfer.sender")
        async let received = fetchTxsByQuery(address: address, queryKey: "transfer.recipient")

        let all = try await sent + received
        return all
            .reduce(into: [String: TransactionResult]()) { dict, tx in
                if dict[tx.txHash] == nil { dict[tx.txHash] = tx }
            }
            .values
            .sorted { $0.height > $1.height }
            .prefix(3)
            .map { $0 }
    }

    @MainActor
    func appendTransaction(_ tx: TransactionResult) {
        sessionManager.appendTransaction(tx)
    }

    @MainActor
    func disconnect() {
        sessionManager.disconnect()
    }

    // MARK: - Transaction Confirmation Polling

    private func awaitTxConfirmation(txHash: String, maxAttempts: Int = 10) async -> TransactionResult? {
        for _ in 0..<maxAttempts {
            try? await Task.sleep(nanoseconds: 1_500_000_000)
            if let result = try? await mobService.getTx(txHash: txHash) {
                return result
            }
        }
        return nil
    }

    // MARK: - REST Transaction History

    private func fetchTxsByQuery(address: String, queryKey: String) async throws -> [TransactionResult] {
        let encodedQuery = "\(queryKey)%3D%27\(address)%27"
        let urlString = "\(Constants.restUrl)cosmos/tx/v1beta1/txs?query=\(encodedQuery)&order_by=ORDER_BY_DESC&pagination.limit=3"
        guard let url = URL(string: urlString) else {
            NSLog("[Repo] fetchTxsByQuery: invalid URL for %@", queryKey)
            return []
        }

        let (data, _) = try await URLSession.shared.data(from: URL(string: urlString)!)

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let txResponses = json["tx_responses"] as? [[String: Any]] else {
            NSLog("[Repo] fetchTxsByQuery %@: failed to parse JSON, %d bytes", queryKey, data.count)
            return []
        }
        NSLog("[Repo] fetchTxsByQuery %@: %d results", queryKey, txResponses.count)

        let txs = json["txs"] as? [[String: Any]]

        return txResponses.enumerated().compactMap { index, txResponse in
            let txBody = txs?[safe: index]

            let txHash = txResponse["txhash"] as? String ?? ""
            let code = txResponse["code"] as? Int ?? -1
            let gasUsed = txResponse["gas_used"] as? String ?? "0"
            let gasWanted = txResponse["gas_wanted"] as? String ?? "0"
            let height = txResponse["height"] as? String ?? "0"
            let rawLog = txResponse["raw_log"] as? String ?? ""
            let timestamp = txResponse["timestamp"] as? String ?? ""

            // Extract fee
            let feeAmount = (txBody?["auth_info"] as? [String: Any])
                .flatMap { $0["fee"] as? [String: Any] }
                .flatMap { $0["amount"] as? [[String: Any]] }
                .flatMap { $0.first }
                .flatMap { $0["amount"] as? String } ?? "0"

            // Raw cosmos message type — fallback label when there's no token transfer.
            let messages = (txBody?["body"] as? [String: Any])?["messages"] as? [[String: Any]]
            let rawShortType: String = {
                guard let msgs = messages, let first = msgs.first,
                      let typeUrl = first["@type"] as? String else { return "" }
                let shortType = typeUrl.components(separatedBy: ".").last ?? typeUrl
                return msgs.count > 1 ? "\(shortType) +\(msgs.count - 1)" : shortType
            }()

            // Find the token transfer involving this wallet, preferring the SBC leg
            // (a Cash Out can also carry a uxion fee transfer we don't want to label on).
            var transferAmount = ""
            var transferDenom = ""
            var transferRecipient = ""
            var isOutgoing = false
            if let events = txResponse["events"] as? [[String: Any]] {
                var best: (amount: String, denom: String, recipient: String, outgoing: Bool)?
                for event in events {
                    guard (event["type"] as? String) == "transfer",
                          let attrs = event["attributes"] as? [[String: Any]] else { continue }
                    var attrMap = [String: String]()
                    for attr in attrs {
                        if let key = attr["key"] as? String, let value = attr["value"] as? String {
                            attrMap[key] = value
                        }
                    }
                    let sender = attrMap["sender"] ?? ""
                    let recipient = attrMap["recipient"] ?? ""
                    guard sender == address || recipient == address else { continue }
                    // "amount" looks like "1000000factory/xion.../sbc" or "1000uxion".
                    // Split leading digits (amount) from the trailing denom — the SBC
                    // factory denom contains digits, so a digit-only strip is wrong.
                    let raw = (attrMap["amount"] ?? "").components(separatedBy: ",").first ?? ""
                    let digits = raw.prefix { $0.isNumber }
                    let denom = String(raw.dropFirst(digits.count))
                    let candidate = (String(digits), denom, recipient, sender == address)
                    if best == nil ||
                        (denom.lowercased().contains("sbc") &&
                         !(best?.denom.lowercased().contains("sbc") ?? false)) {
                        best = candidate
                    }
                }
                if let b = best {
                    transferAmount = b.amount
                    transferDenom = b.denom
                    transferRecipient = b.recipient
                    isOutgoing = b.outgoing
                }
            }

            // App-aware label: Buy / Cash Out for SBC, Send / Received for XION.
            let txType: String = {
                guard !transferDenom.isEmpty else { return rawShortType }
                if transferDenom.lowercased().contains("sbc") {
                    return isOutgoing ? "Cash Out" : "Buy"
                }
                return isOutgoing ? "Send" : "Received"
            }()

            return TransactionResult(
                txHash: txHash,
                success: code == 0,
                gasUsed: gasUsed,
                gasWanted: gasWanted,
                height: Int64(height) ?? 0,
                rawLog: rawLog,
                timestamp: timestamp,
                fee: feeAmount,
                txType: txType,
                amount: transferAmount,
                amountDenom: transferDenom,
                recipient: transferRecipient
            )
        }
    }

    // MARK: - Grant Recovery

    private func withGrantRecovery<T>(_ block: () async throws -> T) async throws -> T {
        do {
            return try await block()
        } catch {
            let message = error.localizedDescription.lowercased()
            if message.contains("authorization not found") || message.contains("fee allowance not found") {
                await sessionManager.setGrantsActive(false)
                throw GrantExpiredError()
            }
            throw error
        }
    }
}

struct GrantExpiredError: LocalizedError {
    var errorDescription: String? { "Session grants have expired. Please reconnect." }
}

enum RepositoryError: LocalizedError {
    case notConnected

    var errorDescription: String? {
        switch self {
        case .notConnected: return "Wallet not connected"
        }
    }
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}

File: xion-android-demo/app/src/main/java/com/burnt/xiondemo/data/repository/XionRepositoryImpl.kt
package com.burnt.xiondemo.data.repository

import com.burnt.xiondemo.data.datasource.MobDataSource
import com.burnt.xiondemo.data.model.BalanceInfo
import com.burnt.xiondemo.data.model.ConnectionStep
import com.burnt.xiondemo.data.model.GrantExpiredException
import com.burnt.xiondemo.data.model.TransactionResult
import com.burnt.xiondemo.data.model.WalletState
import com.burnt.xiondemo.security.Bip39
import com.burnt.xiondemo.security.SecureStorage
import com.burnt.xiondemo.util.Constants
import com.burnt.xiondemo.util.Result
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import org.json.JSONObject
import uniffi.mob.Coin
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class XionRepositoryImpl @Inject constructor(
    private val mobDataSource: MobDataSource,
    private val secureStorage: SecureStorage,
    private val httpClient: OkHttpClient
) : XionRepository {

    private val _walletState = MutableStateFlow<WalletState>(WalletState.Disconnected)
    override val walletState: StateFlow<WalletState> = _walletState.asStateFlow()

    private val _transactionHistory = MutableStateFlow<List<TransactionResult>>(emptyList())
    override val transactionHistory: StateFlow<List<TransactionResult>> = _transactionHistory.asStateFlow()

    // Temporarily holds the session mnemonic between prepareSessionKey and completeConnection
    private var pendingSessionMnemonic: String? = null
    private var pendingSessionKeyAddress: String? = null

    override suspend fun prepareSessionKey(): Result<String> = Result.runCatching {
        _walletState.value = WalletState.Connecting(ConnectionStep.GENERATING_SESSION_KEY)

        val sessionMnemonic = Bip39.generateMnemonic()
        val sessionKeyAddress = mobDataSource.createSigner(sessionMnemonic)

        pendingSessionMnemonic = sessionMnemonic
        pendingSessionKeyAddress = sessionKeyAddress

        sessionKeyAddress
    }.also { result ->
        if (result.isError) {
            _walletState.value = WalletState.Disconnected
            pendingSessionMnemonic = null
            pendingSessionKeyAddress = null
        }
    }

    override suspend fun completeConnection(metaAccountAddress: String): Result<String> = Result.runCatching {
        val sessionMnemonic = pendingSessionMnemonic
            ?: throw IllegalStateException("No pending session key. Call prepareSessionKey() first.")
        val sessionKeyAddress = pendingSessionKeyAddress
            ?: throw IllegalStateException("No pending session key address.")

        _walletState.value = WalletState.Connecting(ConnectionStep.SETTING_UP_GRANTS)

        val expiresAt = System.currentTimeMillis() / 1000 + Constants.SESSION_GRANT_DURATION_SECONDS

        // Upgrade client to session mode — granter/feeGranter are now handled internally
        mobDataSource.upgradeToSessionClient(
            metaAccountAddress = metaAccountAddress,
            treasuryAddress = Constants.TREASURY_ADDRESS,
            sessionExpiresAt = expiresAt
        )

        secureStorage.saveSessionData(
            sessionMnemonic = sessionMnemonic,
            metaAccountAddress = metaAccountAddress,
            sessionKeyAddress = sessionKeyAddress,
            treasuryAddress = Constants.TREASURY_ADDRESS
        )
        secureStorage.saveSessionExpiry(expiresAt)

        _walletState.value = WalletState.Connected(
            metaAccountAddress = metaAccountAddress,
            sessionKeyAddress = sessionKeyAddress,
            treasuryAddress = Constants.TREASURY_ADDRESS,
            grantsActive = true,
            sessionExpiresAt = expiresAt
        )

        // Clear pending state
        pendingSessionMnemonic = null
        pendingSessionKeyAddress = null

        metaAccountAddress
    }.also { result ->
        if (result.isError) {
            _walletState.value = WalletState.Disconnected
            pendingSessionMnemonic = null
            pendingSessionKeyAddress = null
        }
    }

    override suspend fun restoreSession(): Result<Boolean> = Result.runCatching {
        val sessionMnemonic = secureStorage.getSessionMnemonic() ?: return@runCatching false
        val metaAccountAddress = secureStorage.getMetaAccountAddress() ?: return@runCatching false
        val sessionKeyAddress = secureStorage.getSessionKeyAddress() ?: return@runCatching false
        val treasuryAddress = secureStorage.getTreasuryAddress() ?: return@runCatching false
        val expiresAt = secureStorage.getSessionExpiry()

        val now = System.currentTimeMillis() / 1000
        if (expiresAt > 0 && now >= expiresAt) {
            secureStorage.clearAll()
            return@runCatching false
        }

        _walletState.value = WalletState.Connecting(ConnectionStep.GENERATING_SESSION_KEY)

        mobDataSource.createSigner(sessionMnemonic)

        // Upgrade to session client so granter/feeGranter are handled internally
        mobDataSource.upgradeToSessionClient(
            metaAccountAddress = metaAccountAddress,
            treasuryAddress = treasuryAddress,
            sessionExpiresAt = expiresAt
        )

        _walletState.value = WalletState.Connected(
            metaAccountAddress = metaAccountAddress,
            sessionKeyAddress = sessionKeyAddress,
            treasuryAddress = treasuryAddress,
            grantsActive = true,
            sessionExpiresAt = expiresAt
        )

        true
    }.also { result ->
        if (result.isError) {
            _walletState.value = WalletState.Disconnected
        }
    }

    override suspend fun getBalance(): Result<BalanceInfo> = Result.runCatching {
        val state = _walletState.value as? WalletState.Connected
            ?: throw IllegalStateException("Wallet not connected")
        mobDataSource.getBalance(state.metaAccountAddress, Constants.COIN_DENOM)
    }

    override suspend fun getSbcBalance(): Result<BalanceInfo> = Result.runCatching {
        val state = _walletState.value as? WalletState.Connected
            ?: throw IllegalStateException("Wallet not connected")
        mobDataSource.getBalance(state.metaAccountAddress, Constants.BRALE_SBC_ON_CHAIN_DENOM)
    }

    override suspend fun getBlockHeight(): Result<Long> = Result.runCatching {
        mobDataSource.getHeight()
    }

    override suspend fun send(
        toAddress: String,
        amount: String,
        memo: String,
        denom: String
    ): Result<TransactionResult> = withGrantRecovery {
        _walletState.value as? WalletState.Connected
            ?: throw IllegalStateException("Wallet not connected")
        val coins = listOf(Coin(denom = denom, amount = amount))
        val result = mobDataSource.send(
            toAddress = toAddress,
            coins = coins,
            memo = memo.ifBlank { null }
        )
        val confirmed = if (result.success) {
            awaitTxConfirmation(result.txHash) ?: result
        } else {
            result
        }
        _transactionHistory.value = _transactionHistory.value + confirmed
        confirmed
    }

    override suspend fun executeContract(
        contractAddress: String,
        msg: String,
        funds: String?
    ): Result<TransactionResult> = withGrantRecovery {
        _walletState.value as? WalletState.Connected
            ?: throw IllegalStateException("Wallet not connected")
        val fundsCoins = funds?.let { listOf(Coin(denom = Constants.COIN_DENOM, amount = it)) } ?: emptyList()
        val result = mobDataSource.executeContract(
            contractAddress = contractAddress,
            msg = msg.toByteArray(),
            funds = fundsCoins,
            memo = null
        )
        val confirmed = if (result.success) {
            awaitTxConfirmation(result.txHash) ?: result
        } else {
            result
        }
        _transactionHistory.value = _transactionHistory.value + confirmed
        confirmed
    }

    override suspend fun getVaultBalance(): Result<BalanceInfo> = Result.runCatching {
        val state = _walletState.value as? WalletState.Connected
            ?: throw IllegalStateException("Wallet not connected")
        val queryMsg = """{"balance":{"address":"${state.metaAccountAddress}"}}"""
        val responseBytes = mobDataSource.queryContractSmart(
            Constants.VAULT_CONTRACT_ADDRESS,
            queryMsg.toByteArray()
        )
        val responseStr = String(responseBytes)
        val json = org.json.JSONObject(responseStr)
        val coins = json.getJSONArray("coins")
        if (coins.length() == 0) {
            BalanceInfo(amount = "0", denom = Constants.COIN_DENOM)
        } else {
            val firstCoin = coins.getJSONObject(0)
            BalanceInfo(amount = firstCoin.getString("amount"), denom = firstCoin.getString("denom"))
        }
    }

    override suspend fun vaultDeposit(amount: String, denom: String): Result<TransactionResult> = withGrantRecovery {
        _walletState.value as? WalletState.Connected
            ?: throw IllegalStateException("Wallet not connected")
        val msg = """{"deposit":{}}"""
        val fundsCoins = listOf(Coin(denom = denom, amount = amount))
        val result = mobDataSource.executeContract(
            contractAddress = Constants.VAULT_CONTRACT_ADDRESS,
            msg = msg.toByteArray(),
            funds = fundsCoins,
            memo = null
        )
        val confirmed = if (result.success) {
            awaitTxConfirmation(result.txHash) ?: result
        } else {
            result
        }
        _transactionHistory.value = _transactionHistory.value + confirmed
        confirmed
    }

    override suspend fun vaultWithdraw(amount: String, denom: String): Result<TransactionResult> = withGrantRecovery {
        _walletState.value as? WalletState.Connected
            ?: throw IllegalStateException("Wallet not connected")
        val msg = """{"withdraw":{"coins":[{"denom":"$denom","amount":"$amount"}]}}"""
        val result = mobDataSource.executeContract(
            contractAddress = Constants.VAULT_CONTRACT_ADDRESS,
            msg = msg.toByteArray(),
            funds = emptyList(),
            memo = null
        )
        val confirmed = if (result.success) {
            awaitTxConfirmation(result.txHash) ?: result
        } else {
            result
        }
        _transactionHistory.value = _transactionHistory.value + confirmed
        confirmed
    }

    override suspend fun vaultWithdrawAll(): Result<TransactionResult> = withGrantRecovery {
        _walletState.value as? WalletState.Connected
            ?: throw IllegalStateException("Wallet not connected")
        val msg = """{"withdraw_all":{}}"""
        val result = mobDataSource.executeContract(
            contractAddress = Constants.VAULT_CONTRACT_ADDRESS,
            msg = msg.toByteArray(),
            funds = emptyList(),
            memo = null
        )
        val confirmed = if (result.success) {
            awaitTxConfirmation(result.txHash) ?: result
        } else {
            result
        }
        _transactionHistory.value = _transactionHistory.value + confirmed
        confirmed
    }

    override suspend fun getTx(txHash: String): Result<TransactionResult> = Result.runCatching {
        mobDataSource.getTx(txHash)
    }

    override suspend fun getRecentTransactions(address: String): Result<List<TransactionResult>> = Result.runCatching {
        withContext(Dispatchers.IO) {
            val sent = fetchTxsByQuery(address, "transfer.sender%3D%27${address}%27")
            val received = fetchTxsByQuery(address, "transfer.recipient%3D%27${address}%27")
            val merged = (sent + received)
                .distinctBy { it.txHash }
                .sortedByDescending { it.height }
                .take(3)
            merged
        }
    }

    private fun fetchTxsByQuery(userAddress: String, queryValue: String): List<TransactionResult> {
        val url = "${Constants.REST_URL}cosmos/tx/v1beta1/txs" +
            "?query=${queryValue}" +
            "&order_by=ORDER_BY_DESC" +
            "&pagination.limit=3"
        val request = okhttp3.Request.Builder().url(url).get().build()
        val response = httpClient.newCall(request).execute()
        return response.use { resp ->
            if (!resp.isSuccessful) return@use emptyList()
            val body = resp.body?.string() ?: return@use emptyList()
            val json = JSONObject(body)
            val txResponses = json.optJSONArray("tx_responses") ?: return@use emptyList()
            val txs = json.optJSONArray("txs")
            (0 until txResponses.length()).map { i ->
                val tx = txResponses.getJSONObject(i)
                val txBody = txs?.optJSONObject(i)

                // Extract fee from tx body
                val feeAmount = txBody?.optJSONObject("auth_info")
                    ?.optJSONObject("fee")
                    ?.optJSONArray("amount")
                    ?.optJSONObject(0)
                    ?.optString("amount", "0") ?: "0"

                // Extract tx type from messages
                val messages = txBody?.optJSONObject("body")
                    ?.optJSONArray("messages")
                val txType = if (messages != null && messages.length() > 0) {
                    val typeUrl = messages.getJSONObject(0).optString("@type", "")
                    val shortType = typeUrl.substringAfterLast(".")
                    if (messages.length() > 1) "$shortType +${messages.length() - 1}" else shortType
                } else ""

                // Extract amount, denom, and recipient from transfer events
                var transferAmount = ""
                var transferDenom = ""
                var transferRecipient = ""
                var transferSender = ""
                val events = tx.optJSONArray("events")
                if (events != null) {
                    for (e in 0 until events.length()) {
                        val event = events.getJSONObject(e)
                        if (event.optString("type") == "transfer") {
                            val attrs = event.optJSONArray("attributes") ?: continue
                            val attrMap = mutableMapOf<String, String>()
                            for (a in 0 until attrs.length()) {
                                val attr = attrs.getJSONObject(a)
                                attrMap[attr.optString("key")] = attr.optString("value")
                            }
                            val sender = attrMap["sender"] ?: ""
                            val recipient = attrMap["recipient"] ?: ""
                            val rawAmount = attrMap["amount"] ?: ""
                            if (sender == userAddress || recipient == userAddress) {
                                // Parse "1200000factory/xion.../sbc" or "5000000uxion"
                                // Extract leading digits as amount, rest as denom
                                val match = Regex("^(\\d+)(.*)$").find(rawAmount)
                                transferAmount = match?.groupValues?.get(1) ?: rawAmount.replace(Regex("[^0-9]"), "")
                                transferDenom = match?.groupValues?.get(2) ?: ""
                                transferRecipient = recipient
                                transferSender = sender
                                break
                            }
                        }
                    }
                }

                // Determine display-friendly tx type
                val displayTxType = when {
                    transferRecipient.startsWith("xion1amma") || transferRecipient.startsWith("xion1yymx") -> "Cash Out"
                    transferSender.startsWith("xion1amma") || transferSender.startsWith("xion1yymx") -> "Buy SBC"
                    transferDenom.contains("sbc", ignoreCase = true) -> "SBC Transfer"
                    txType == "MsgExec" && transferAmount.isNotEmpty() -> "Send"
                    else -> txType
                }

                TransactionResult(
                    txHash = tx.optString("txhash", ""),
                    success = tx.optInt("code", -1) == 0,
                    gasUsed = tx.optString("gas_used", "0"),
                    gasWanted = tx.optString("gas_wanted", "0"),
                    height = tx.optLong("height", 0),
                    rawLog = tx.optString("raw_log", ""),
                    timestamp = tx.optString("timestamp", ""),
                    fee = feeAmount,
                    txType = displayTxType,
                    amount = transferAmount,
                    amountDenom = if (transferDenom.contains("sbc", ignoreCase = true)) "SBC" else "XION",
                    recipient = transferRecipient
                )
            }
        }
    }

    override fun appendTransaction(tx: TransactionResult) {
        _transactionHistory.value = _transactionHistory.value + tx
    }

    override fun disconnect() {
        mobDataSource.disconnect()
        secureStorage.clearAll()
        _walletState.value = WalletState.Disconnected
        _transactionHistory.value = emptyList()
        pendingSessionMnemonic = null
        pendingSessionKeyAddress = null
    }

    private suspend fun awaitTxConfirmation(
        txHash: String,
        maxAttempts: Int = 10,
        delayMs: Long = 1500
    ): TransactionResult? {
        repeat(maxAttempts) {
            delay(delayMs)
            try {
                return mobDataSource.getTx(txHash)
            } catch (_: Exception) {
                // tx not yet included in a block, retry
            }
        }
        return null
    }

    private suspend fun <T> withGrantRecovery(block: suspend () -> T): Result<T> = Result.runCatching {
        try {
            block()
        } catch (e: Exception) {
            val message = e.message?.lowercase() ?: ""
            if (message.contains("authorization not found") || message.contains("fee allowance not found")) {
                val state = _walletState.value
                if (state is WalletState.Connected) {
                    _walletState.value = state.copy(grantsActive = false)
                }
                throw GrantExpiredException("Session grants have expired. Please reconnect.")
            }
            throw e
        }
    }
}
